lib/icinga/dependency-group.cpp

```diff
@@ -0,0 +1,384 @@
+ * of the group. A dependency group is considered unreachable when none of the parent Checkables is reachable. However,
+ * a dependency group may still be marked as failed even when it has reachable parent Checkables, but an unreachable
+ * group has always a failed state.
+ *
+ * @param child The child Checkable to evaluate the state for.
+ * @param dt The dependency type to evaluate the state for, defaults to DependencyState.
+ * @param rstack The recursion stack level to prevent infinite recursion, defaults to 0.
+ *
+ * @return - Returns the state of the current dependency group.
+ */
+DependencyGroup::State DependencyGroup::GetState(const Checkable* child, DependencyType dt, int rstack) const
+{
+	MembersMap members;
+	{
+		// We don't want to hold the mutex lock for the entire evaluation, thus we just need to operate on a copy.
+		std::lock_guard lock(m_Mutex);
+		members = m_Members;
+	}
+
+	size_t reachable = 0, available = 0;
+
+	for (auto& [_, dependencies] : members) {
+		ASSERT(!dependencies.empty());
+
+		auto [begin, end] = dependencies.equal_range(child);
+		for (auto it(begin); it != end; ++it) {
+			Dependency::Ptr dependency(it->second);
+			if (dependency->GetParent()->IsReachable(dt, rstack)) {
+				reachable++;
+
+				// Only reachable parents are considered for availability. If they are unreachable and checks are disabled,
+				// they could be incorrectly treated as available otherwise.
+				if (dependency->IsAvailable(dt)) {
+					available++;
+				}
+			}
+
+			if (dt == DependencyState) {
+				// Dependencies are grouped by parent and all config attributes affecting their availability.
+				// However, their disable_{checks,notifications} attrs aren't taken into account when grouping them,
+				// thus the map may contain multiple dependencies for that child with different config of these attrs.
+				// Therefore, if this state evaluation is for DependencyState, we can break here as we only need to
+				// consider only a single dependency as they all share the same state, otherwise we need to evaluate
+				// all of them.
+				break;
+			}
+		}
+	}
+
+	if (IsRedundancyGroup()) {
+		// The state of a redundancy group is determined by the best state of any parent. If any parent ist reachable,
+		// the redundancy group is reachable, analogously for availability. Note that an unreachable group cannot be
+		// available as reachable = 0 implies available = 0.
+		return {reachable > 0, available > 0};
+	}
+
+	// For dependencies without a redundancy group, members.size() will be 1 in almost all cases. It will only
+	// contain more elements if there are duplicate dependency config objects between two checkables. In this case,
+	// all of them have to be reachable or available as they don't provide redundancy. Note that unreachable implies
+	// unavailable here as well as only reachable parents count towards the number of available parents.
+	return {reachable == members.size(), available == members.size()};
```

Comment:
`members.size()` will no longer work here if you may check (and thus count) more than one dependency per entry.